internal/service/vmservice/bootstrap_test.go

```diff
@@ -109,11 +111,65 @@ func TestReconcileBootstrapData_UpdateStatus(t *testing.T) {
 	require.True(t, *machineScope.ProxmoxMachine.Status.BootstrapDataProvided)
 }
 
+func TestReconcileBootstrapData_BadInjector(t *testing.T) {
+	machineScope, _, kubeClient := setupReconcilerTest(t)
+	vm := newVMWithNets("virtio=A6:23:64:4D:84:CB,bridge=vmbr0")
+	machineScope.SetVirtualMachine(vm)
+	machineScope.ProxmoxMachine.Status.IPAddresses = map[string]infrav1alpha1.IPAddress{infrav1alpha1.DefaultNetworkDevice: {IPV4: "10.10.10.10"}}
+	createIP4AddressResource(t, kubeClient, machineScope, infrav1alpha1.DefaultNetworkDevice, "10.10.10.10")
+	createBootstrapSecret(t, kubeClient, machineScope)
+
+	getISOInjector = func(_ *proxmox.VirtualMachine, _ []byte, _, _ cloudinit.Renderer) isoInjector {
+		return FakeISOInjector{Error: errors.New("bad FakeISOInjector")}
+	}
+	t.Cleanup(func() { getISOInjector = defaultISOInjector })
+
+	requeue, err := reconcileBootstrapData(context.Background(), machineScope)
+	require.Error(t, err)
+	require.Equal(t, err.Error(), "cloud-init iso inject failed: bad FakeISOInjector")
+	require.False(t, requeue)
+	require.True(t, conditions.Has(machineScope.ProxmoxMachine, infrav1alpha1.VMProvisionedCondition))
+	require.Nil(t, machineScope.ProxmoxMachine.Status.BootstrapDataProvided)
+}
+
 func TestGetBootstrapData_MissingSecretName(t *testing.T) {
 	machineScope, _, _ := setupReconcilerTest(t)
 
 	data, err := getBootstrapData(context.Background(), machineScope)
 	require.Error(t, err)
+	require.Equal(t, err.Error(), "machine has no bootstrap data")
+	require.Nil(t, data)
+}
+
+func TestGetBootstrapData_MissingSecretNotName(t *testing.T) {
```

Comment:
isn't this similar to `estGetBootstrapData_MissingSecretName`

Reply:
It is similar but it tests a different codepath: `TestGetBootstrapData_MissingSecretName` tests the `if scope.Machine.Spec.Bootstrap.DataSecretName == nil {` whereas my new test `TestGetBootstrapData_MissingSecretNotName` tests the  `if err := scope.GetBootstrapSecret(ctx, secret); err != nil {` that follows.